app/module/module.go

```diff
@@ -0,0 +1,353 @@
+		Validators: validatorUpdates,
+	}
+}
+
+// ExportGenesis performs export genesis functionality for modules
+func (m *Manager) ExportGenesis(ctx sdk.Context, cdc codec.JSONCodec, version uint64) map[string]json.RawMessage {
+	genesisData := make(map[string]json.RawMessage)
+	modules := m.versionedModules[version]
+	for _, moduleName := range m.OrderExportGenesis {
+		genesisData[moduleName] = modules[moduleName].ExportGenesis(ctx, cdc)
+	}
+
+	return genesisData
+}
+
+// assertNoForgottenModules checks that we didn't forget any modules in the
+// SetOrder* functions.
+func (m *Manager) assertNoForgottenModules(setOrderFnName string, moduleNames []string) {
+	ms := make(map[string]bool)
+	for _, m := range moduleNames {
+		ms[m] = true
+	}
+	var missing []string
+	for _, m := range m.allModules {
+		if _, ok := ms[m.Name()]; !ok {
+			missing = append(missing, m.Name())
+		}
+	}
+	if len(missing) != 0 {
+		panic(fmt.Sprintf(
+			"%s: all modules must be defined when setting %s, missing: %v", setOrderFnName, setOrderFnName, missing))
+	}
+}
+
+// MigrationHandler is the migration function that each module registers.
+type MigrationHandler func(sdk.Context) error
+
+// VersionMap is a map of moduleName -> version
+type VersionMap map[string]uint64
+
+// RunMigrations performs in-place store migrations for all modules. This
+// function MUST be called when the state machine changes appVersion
+func (m Manager) RunMigrations(ctx sdk.Context, cfg sdkmodule.Configurator, fromVersion, toVersion uint64) error {
+	c, ok := cfg.(configurator)
+	if !ok {
+		return sdkerrors.ErrInvalidType.Wrapf("expected %T, got %T", configurator{}, cfg)
+	}
+	modules := m.OrderMigrations
+	if modules == nil {
+		modules = DefaultMigrationsOrder(m.ModuleNames(toVersion))
+	}
+	currentVersionModules, exists := m.versionedModules[fromVersion]
+	if !exists {
+		return sdkerrors.ErrInvalidVersion.Wrapf("fromVersion %d not supported", fromVersion)
+	}
+	nextVersionModules, exists := m.versionedModules[toVersion]
+	if !exists {
+		return sdkerrors.ErrInvalidVersion.Wrapf("toVersion %d not supported", toVersion)
+	}
+
+	for _, moduleName := range modules {
+		_, currentModuleExists := currentVersionModules[moduleName]
+		nextModule, nextModuleExists := nextVersionModules[moduleName]
+
+		// if the module exists for both upgrades
+		if currentModuleExists && nextModuleExists {
+			err := c.runModuleMigrations(ctx, moduleName, fromVersion, toVersion)
+			if err != nil {
+				return err
+			}
+		} else if !currentModuleExists && nextModuleExists {
+			ctx.Logger().Info(fmt.Sprintf("adding a new module: %s", moduleName))
+			moduleValUpdates := nextModule.InitGenesis(ctx, c.cdc, nextModule.DefaultGenesis(c.cdc))
+			// The module manager assumes only one module will update the
+			// validator set, and it can't be a new module.
+			if len(moduleValUpdates) > 0 {
+				return sdkerrors.ErrLogic.Wrap("validator InitGenesis update is already set by another module")
```

Comment:
[Question] In the `EndBlock` , under a similar circumstances, where there are multiple validator updates, the method panics, but here it returns an error. Just wondering whether the two should behave similarly or not.

Reply:
This section was taken from the original module manager in the cosmos sdk